### ISSUE TITLE
Offline map and themes folder incorrect after re-install and restore of backup

### ISSUE THREAD

**Issue by MagpieFourtyTwo** [cgeo/cgeo#10690]
## Bug description 
Had to re-install the app and restored caches and settings from latest backup. Went through the assistant, of course, who even warned me on restore of the settings that the offline maps folder will be changed to ``SDcard/Locus/mapsVector`` (he even asked for access permission), but in the app there are no offline maps available at all and the setting ``Folder with Offline Maps`` still says ``(internal)/cgeo/maps``. The setting of ``Map Themes Folder`` is ``(internal)/cgeo/maps/_themes``, thus wrong, too.

## Reproduce
- perform a backup with settings and passwords
- uninstall, re-install and start
- Msg ``Welcome ...`` => [Next] => Msg ``App needs permissions`` => [Next]
- Storage permission => [Allow]
- Location permission => [Allow while using]
- Set folder: Base => ``(internal)/cgeo`` => [Allow]
- Services => [Skip] (should be restored by backup restore)
- Advanced Config => [Restore]
    [X] caches and waypoints
    [X] program settings
    ... Restore ...
- Msg: ``Folder "Offline Maps" will be changed to "SDcard/Locus/mapsVector"`` => Re-Grant Access => Ok => Allow0 fiole
- Msg ``Change Offline Map Folder`` => "... current folder contains 0 files, 0 folders and has a size of 0 B ...
    Select folder and ... [X] No further action"
- Done => Finish

### Actual result after these steps 
No offline maps available, setting ``Folder with offline maps`` still points to ``internal/cgeo/maps``.
Same applies to ``Map Themes Folder``, pointing to ``internal/cgeo/maps/_themes``.

### Expected result after these steps 
Offline maps should be available,
setting ``Folder with offline maps`` should point to ``SDcard/Locus/mapsVector`` (as it was in backup).
setting ``Map Themes Folder`` should point to ``SDcard/Locus/mapsVector/_themes/Elevate4`` (as it was in backup).

## c:geo version 
2021.05.20-NB1-730260e

## Reproducible 
Yes

## System information
```
--- System information ---
c:geo version: 2021.05.20-NB1-730260e

Device:
-------
- Device type: VOG-L29 (VOG-L29EEA, HUAWEI)
- Available processors: 8
- Android version: 10
- Android build: VOG-L29 10.1.0.150(C431E19R2P5)
- Sailfish OS detected: false
- Google Play services: disabled - 21.15.15 (120400-371058782)
- HW acceleration: enabled (default state)

Sensor and location:
-------
- Low power mode: inactive
- Compass capabilities: yes
- Rotation vector sensor: present
- Orientation sensor: present
- Magnetometer & Accelerometer sensor: present
- Direction sensor used: rotation vector

Program settings:
-------
- Hide caches: own/found archived
- Hide waypoints: -
- Set language: en_DE (system default)
- System date format: dd/MM/y
- Debug mode active: no
- Live map mode: true
- OSM multi-threading: true / threads: 4
- Global filter: display all caches
- Last backup: 19 May, 23:48
- Routing mode: Car
 -Settings: v5, Count:262
- Map: Google: Map
  - Id: cgeo.geocaching.maps.google.v2.GoogleMapProvider$GoogleMapSource
  - Atts: none
  - Theme: Elevate F160 L120 S130 DS.xml

Services:
-------
- Geocaching sites enabled:
   geocaching.com: Logged in (Login OK) / PREMIUM
   extremcaching.com: Logged in (Login OK)
- Geocaching.com date format: dd MMM yy
- Routing: external / BRouter installed: true
- Installed c:geo plugins: contacts

Permissions & paths:
-------
- Fine location permission: granted
- Write external storage permission: granted
- System internal c:geo dir: /data/user/0/cgeo.geocaching (61,0 GB free) v2 internal isDir(7 entries)
- Legacy User storage c:geo dir: /storage/emulated/0/cgeo (60,9 GB free) v2 external non-removable isDir(7 entries)
- Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (60,9 GB free) v2 external non-removable isDir(1 entries)
- Internal theme sync (is turned ON): /data/user/0/cgeo.geocaching/MapThemeData (61,0 GB free) v2 internal isDir(8 entries)
- Public Folders: #11
  - BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=10, dirs:>=13, totalFileSize:>=1,1 GB, free space: 60,9 GB, files on device: 13278715)
  - OFFLINE_MAPS: ‎‏‎‎‎‎‎‏‎‏‏‏‎‎‎‎‎‎‏‎‎‏‎‎‎‎‏‏‏‏‏‏‎‏‎‏‏‎‏‏‎‎‎‏‏‏‏‎‏‏‎‎‎‎‎‏‎‎‏‏‎‎‎‎‏‏‏‏‎‏‎‎‎‏‏‏‎‎‏‏‏‎‏‎‏‎‎‏‎‎‏‏‎‎‏‎SD card‎‏‎‎‏‎/Locus/mapsVector (User-Defined)[‎‏‎‎‎‎‎‏‎‏‏‏‎‎‎‎‎‎‏‎‎‏‎‎‎‎‏‏‏‏‏‏‎‏‎‏‏‎‏‏‎‎‎‏‏‏‏‎‏‏‎‎‎‎‎‏‎‎‏‏‎‎‎‎‏‏‏‏‎‏‎‎‎‏‏‏‎‎‏‏‏‎‏‎‏‎‎‏‎‎‏‏‎‎‏‎SD card‎‏‎‎‏‎/Locus/mapsVector[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/0123-4567%3ALocus%2FmapsVector::]] (Uri: content://com.android.externalstorage.documents/tree/0123-4567%3ALocus%2FmapsVector/document/0123-4567%3ALocus%2FmapsVector, Av:true, files:>=46, dirs:>=8, totalFileSize:>=20,7 GB, free space: 154,7 GB, files on device: 0)
  - OFFLINE_MAP_THEMES: ‎‏‎‎‎‎‎‏‎‏‏‏‎‎‎‎‎‎‏‎‎‏‎‎‎‎‏‏‏‏‏‏‎‏‎‏‏‎‏‏‎‎‎‏‏‏‏‎‏‏‎‎‎‎‎‏‎‎‏‏‎‎‎‎‏‏‏‏‎‏‎‎‎‏‏‏‎‎‏‏‏‎‏‎‏‎‎‏‎‎‏‏‎‎‏‎SD card‎‏‎‎‏‎/Locus/mapsVector/_themes/Elevate4 (User-Defined)[‎‏‎‎‎‎‎‏‎‏‏‏‎‎‎‎‎‎‏‎‎‏‎‎‎‎‏‏‏‏‏‏‎‏‎‏‏‎‏‏‎‎‎‏‏‏‏‎‏‏‎‎‎‎‎‏‎‎‏‏‎‎‎‎‏‏‏‏‎‏‎‎‎‏‏‏‎‎‏‏‏‎‏‎‏‎‎‏‎‎‏‏‎‎‏‎SD card‎‏‎‎‏‎/Locus/mapsVector/_themes/Elevate4[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/0123-4567%3ALocus%2FmapsVector%2F_themes%2FElevate4::]] (Uri: content://com.android.externalstorage.documents/tree/0123-4567%3ALocus%2FmapsVector%2F_themes%2FElevate4/document/0123-4567%3ALocus%2FmapsVector%2F_themes%2FElevate4, Av:true, files:418, dirs:1, totalFileSize:3,4 MB, free space: 154,7 GB, files on device: 0)
  - LOGFILES: /cgeo/logfiles (Default)[/cgeo/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/logfiles]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Flogfiles, Av:true, files:1, dirs:0, totalFileSize:119 B, free space: 60,9 GB, files on device: 13278715)
  - GPX: /cgeo/gpx (Default)[/cgeo/gpx[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/gpx]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fgpx, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 60,9 GB, files on device: 13278715)
  - BACKUP: /cgeo/backup (Default)[/cgeo/backup[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/backup]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fbackup, Av:true, files:>=15, dirs:>=6, totalFileSize:>=1,8 GB, free space: 60,9 GB, files on device: 13278715)
  - FIELD_NOTES: /cgeo/field-notes (Default)[/cgeo/field-notes[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/field-notes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Ffield-notes, Av:true, files:50, dirs:0, totalFileSize:189,7 KB, free space: 60,9 GB, files on device: 13278715)
  - SPOILER_IMAGES: /cgeo/GeocachePhotos (Default)[/cgeo/GeocachePhotos[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/GeocachePhotos]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2FGeocachePhotos, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 60,9 GB, files on device: 13278715)
  - ROUTING_BASE: /cgeo/routing (Default)[/cgeo/routing[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting, Av:true, files:15, dirs:1, totalFileSize:574,6 MB, free space: 60,9 GB, files on device: 13278715)
  - ROUTING_TILES: /cgeo/routing/segments4 (Default)[/cgeo/routing/segments4[PERSISTABLE_FOLDER(ROUTING_BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing/segments4]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting%2Fsegments4, Av:true, files:5, dirs:0, totalFileSize:574,5 MB, free space: 60,9 GB, files on device: 13278715)
  - TEST_FOLDER: [Legacy]/data/user/0/cgeo.geocaching/files/unittest (Default)[/data/user/0/cgeo.geocaching/files/unittest[FILE#1:p-file:///data/user/0/cgeo.geocaching/files::/unittest]] (Uri: file:///data/user/0/cgeo.geocaching/files/unittest, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 61,0 GB, files on device: -1)
- Map render theme path: Elevate F160 L120 S130 DS.xml
- PersistedDocumentUris: #1
- TRACK: null
- Persisted Uri Permissions: #3
  - content://com.android.externalstorage.documents/tree/0123-4567%3ALocus%2FmapsVector%2F_themes%2FElevate4 (20 May, 13:21):RW
  - content://com.android.externalstorage.documents/tree/primary%3Acgeo (20 May, 13:10):RW
  - content://com.android.externalstorage.documents/tree/0123-4567%3ALocus%2FmapsVector (20 May, 13:21):RW
- Database: /data/user/0/cgeo.geocaching/databases/data (v95, Size:358,6 MB) on system internal storage
--- End of system information ---
```

## Additional Information
IIRC, GPX-folder was set to ``(internal)/Download`` and routing to external BRouter, but now (i. e. after restore of the backup) it's set to ``(internal)/cgeo/GPX`` and internal routing ... but in fact I'm not completely sure about these, I may have changed them while testing ... would nevertheless be great if the assistant's handling of these two would be checked, too. Will test later on ...

**Comment by eddiemuc** [comment]
Sadly, I am almost certain that the "loop of regrant" does not work any more after I changed CSAH to be state-aware. Always the tricky things which fail....

@MagpieFourtyTwo if I understand you correctly, you set at least two folders to individual values in before creating the backup (maps and themes, maybe even more e.g. gpx, brouter), but during restore routine you were only asked to regrant access for one of them (maps). Correct? This would be an indication for me that the "regrant loop" does not work correctly.

**Comment by eddiemuc** [comment]
@MagpieFourtyTwo tried to recreate the error, but restore works for me -> except for restoring the TrackUtils-file (which fails with same error as described in #10684). But something else occured to me: after regranting access for folder "maps", the system should have asked you whether you want to "move/copy" the data in the folders. Since this did not happen for your I assume another error.

It would be helpful to get a debug logfile for this situation. Are you able to reproduce the situation and provide such a log?

**Comment by MagpieFourtyTwo** [comment]
> you set at least two folders to individual values in before creating the backup (maps and themes, maybe even more e.g. gpx, brouter), but during restore routine you were only asked to regrant access for one of them (maps). Correct?

Correct.

> But something else occured to me: after regranting access for folder "maps", the system should have asked you whether you want to "move/copy" the data in the folders. Since this did not happen for your I assume another error.

Sorry for inadvertently laying out a false hint and misleading you, simply forgot to mention this step - just added it in the OP.
Nevertheless, after selecting one ``[X] Do nothing`` there were no more dialogues anymore. Wondered for a second, as I would have expected him to ask for the themes folder, too - but as this is just a subfolder of the maps folder, I thought, he has already got permission for the parent anyway, so everything will be fine ...

> It would be helpful to get a debug logfile for this situation. Are you able to reproduce the situation and provide such a log?

No problem. Already re-installed twice today, and all good things come in threes, anyway.  ;)
Just let me know how to activate debugging *before* installing c:geo ... by placing ``log.properties.txt`` in the logfiles folder?
Still have one with the following content:

```
logging.minlevel=VERBOSE
logging.mincallerinfolevel=VERBOSE
logging.throwonerror=false
logging.callerinfomaxdepth=10
```

BTW: Exported a GPX this afternoon, which was written to ``(internal)/cgeo/GPX`` and now I'm absolutely sure that my GPX folder was ``(internal)/Download`` before, i. e. restore has missed this setting, too, cause these exports always went to Download - cause there are all the other files I previously exported. In fact there are is a bunch of GPX files and the newest is dated May, 16th, thus this sunday, when I exported the last one.

**Comment by MagpieFourtyTwo** [comment]
Just seen: Although I used c:geo again since the restore, the value of ``Settings`` => ``Map`` => ``Select Map Source`` currently is empty, which looks strange to me - a) this should be set by restore, too, I would guess, and b) isn't this at least set when selecting a map in Live Map view ...?

**Comment by MagpieFourtyTwo** [comment]
... and in ``Settings`` => ``System`` => ``Local File System`` the value of ``Base folder for public local data`` is set to ``/cgeo (User defined)`` - isn't this the default?  Even on restore I just accepted the default ...

Ah, if I start to change it, c:geo says that ``[Legacy]/storage/emulated/0/cgeo`` is the default ... seems to be nothing than an inconsistency in interpretation ...?

**Comment by MagpieFourtyTwo** [comment]
Now thing get weird ... ok, step after step:
- re-installed c:geo (first try with just the ``log.properties.txt`` file activated, but output was not really meaningful)
- thus re-installed again, but this time skipped the assistant on first go (but had to grant access to files and location)
- then set ``Activate debug log [X}``, went back to Android's home screen and opened c:geo again
   (to get back to the assistant, as I did not know where to start him manually)
- and on this second start I followed the assistant ...
- did all steps like described in the OP, but this time ... he not just asked for the themes, but even for the GPX folder
- and at the very end c:geo said it has to restart to make the changes effective.

BTW: This very last restart message will *not* be shown, either, if I immediately follow the assistant in the first go.

As the second go worked as it should, I guess it's logfile is pretty superfluous.
So I just add the [logcat](https://github.com/cgeo/cgeo/files/6518932/logcat-2021-05-20-23-33-09-2.txt) of the first try (i. e. when following the assistant without previously setting ``Activate debug log``)

If you need more/different stuff, feel free to ask.

**Comment by MagpieFourtyTwo** [comment]
Seems as if (my) c:geo needs to be restarted after installation and before starting the assistant ...?

**Comment by MagpieFourtyTwo** [comment]
Or a possible timing problem ...? As logging consumes time, giving routines more time to finish successfully ...?

**Comment by eddiemuc** [comment]
Hi @MagpieFourtyTwo thanks a lot for the huge testing effort! So the problem occurs only immediately after reinstall, I will try this out in emulator. Nothing suspicious in the logfile, guess I have to recreate this myself.

It is not surprising that the log settings are not used on first start after reinstall, because c:geo checks for the logfile at startup and at this time no permission is granted to the logging folder yet. I currently have no idea for an easy way to work around this, so I guess we have to live with that.

**Comment by eddiemuc** [comment]
Does anyone know why the restart is needed after wizard completion anyway?

**Comment by MagpieFourtyTwo** [comment]
> at this time no permission is granted to the logging folder yet

Thought so. Perhaps re-check for ``log.properties.txt`` after storage permission has been granted ...?
Although ... general storage per

> Does anyone know why the restart is needed after wizard completion anyway?

Easy. Windows principle. Reboot is always good.  ;)

**Comment by eddiemuc** [comment]
Got it, `BackupUtils` were not correctly included in `InstallWizardActivity`'s `onActiviyResult`. Fixed with PR #10706

**Comment by MagpieFourtyTwo** [comment]
Fine! Thx! Will re-check on next re-install.  ;)
